internal/testsuite/eventwatcher/eventwatcher.go

```diff
@@ -0,0 +1,266 @@
+				}
+			}
+
+			// Return an error if the job has exited without us seeing all expected events.
+			if isTerminalEvent(actual) && i < len(expected) {
+				return &ErrUnexpectedEvent{
+					expected: expected[i],
+					actual:   actual,
+				}
+			}
+		}
+	}
+}
+
+func isTerminalEvent(msg *api.EventMessage) bool {
+	switch msg.Events.(type) {
+	case *api.EventMessage_Failed:
+		return true
+	case *api.EventMessage_Succeeded:
+		return true
+	case *api.EventMessage_Cancelled:
+		return true
+	}
+	return false
+}
+
+// WithCancelOnNoActiveJobs returns a context that is cancelled when there are no active jobs,
+// or if the parent is cancelled.
+func ErrorOnNoActiveJobs(parent context.Context, C chan *api.EventMessage, jobIds map[string]bool) error {
+	numActive := 0
+	numRemaining := len(jobIds)
+	exitedByJobId := make(map[string]bool)
+	for {
+		select {
+		case <-parent.Done():
+			return nil
+		case msg := <-C:
+			if e := msg.GetSubmitted(); e != nil {
+				numActive++
+			} else if e := msg.GetSucceeded(); e != nil {
+				if _, ok := exitedByJobId[e.JobId]; ok {
+					return errors.Errorf("received multiple terminal events for job %s", e.JobId)
+				}
+				exitedByJobId[e.JobId] = true
+				if _, ok := jobIds[e.JobId]; ok {
+					numRemaining--
+				}
+				numActive--
+			} else if e := msg.GetFailed(); e != nil {
+				if _, ok := exitedByJobId[e.JobId]; ok {
+					return errors.Errorf("received multiple terminal events for job %s", e.JobId)
+				}
+				exitedByJobId[e.JobId] = true
+				if _, ok := jobIds[e.JobId]; ok {
+					numRemaining--
+				}
+				numActive--
+			}
+			if numRemaining <= 0 {
+				return errors.New("all jobs exited")
+			}
+		}
+	}
+}
+
+// WithCancelOnFailed returns a context that is cancelled if a job fails,
+// or if the parent is cancelled.
+func ErrorOnFailed(parent context.Context, C chan *api.EventMessage) error {
+	for {
+		select {
+		case <-parent.Done():
+			return parent.Err()
+		case msg := <-C:
+			if failedEvent := msg.GetFailed(); failedEvent != nil {
+				err := errors.Errorf("job failed")
+				return errors.WithMessagef(err, "%v", failedEvent)
+			}
+		}
+	}
+}
+
+// GetFromIngresses listens for ingressInfo messages and tries to download from each ingress.
+// Returns false if any download fails.
+func GetFromIngresses(parent context.Context, C chan *api.EventMessage) error {
+	g, ctx := errgroup.WithContext(parent)
+	for {
+		select {
+		case <-parent.Done():
+			return g.Wait()
+		case <-ctx.Done(): // errgroup cancelled
+			return g.Wait()
+		case msg := <-C:
+			if ingressInfo := msg.GetIngressInfo(); ingressInfo != nil {
+				for _, host := range ingressInfo.IngressAddresses {
+					ctxWithTimeout, _ := context.WithTimeout(ctx, 10*time.Second)
+					host := host
+					g.Go(func() error { return getFromIngress(ctxWithTimeout, host) })
+				}
+			}
+		}
+	}
+}
+
+func getFromIngress(ctx context.Context, host string) error {
+	tokens := strings.Split(host, ".")
+	if len(tokens) == 0 {
```

Comment:
I'm not sure I follow this- way does taking the last token here yield a valid host?

Reply:
You're right; this is wrong. Fixing it.